packages/paymasters/contracts/PermitERC20UniswapV3Paymaster.sol

```diff
@@ -91,25 +96,88 @@ contract PermitERC20UniswapV3Paymaster is BasePaymaster, ERC2771Recipient {
+        IChainlinkOracle[] memory _priceFeeds,
+        string[] memory _permitMethodSignatures,
+        uint24[] memory _poolFees) public onlyOwner {
+        tokens = _tokens;
+        // allow uniswap to transfer from paymaster balance
+        for (uint256 i = 0; i < tokens.length; i++) {
+            TokenSwapData memory data;
+            IERC20Metadata token = tokens[i];
+            token.approve(address(uniswap), type(uint256).max);
+            data.priceDivisor = 10 ** uint256(_priceFeeds[i].decimals() + token.decimals());
+            data.priceFeed = _priceFeeds[i];
+            data.permitMethodSignature = bytes4(keccak256(bytes(_permitMethodSignatures[i])));
+            data.uniswapPoolFee = _poolFees[i];
+            tokensSwapData[token] = data;
+        }
+    }
+
+    function getTokens() public view returns (IERC20Metadata[] memory){
+        return tokens;
+    }
+
+    function getTokenSwapData(IERC20Metadata token) public view returns (TokenSwapData memory) {
+        return tokensSwapData[token];
+    }
+
+    function refillHubDeposit(uint256 amount) public payable onlyOwner {
+        _refillHubDeposit(amount);
+    }
+
+    function withdrawTokens(IERC20Metadata[] calldata _tokens, address target, uint256[] calldata amounts) public onlyOwner {
+        for (uint256 i = 0; i < _tokens.length; i++) {
+            _tokens[i].safeTransfer(target, amounts[i]);
+        }
+    }
+
+    function _calculateCharge(
+        GsnTypes.RelayData calldata relayData,
+        uint256 gasUsed,
+        uint256 priceQuote,
+        uint256 priceDivisor
+    ) internal
     view
-    returns (uint256 tokenPreCharge) {
-        uint256 ethMaxCharge =
-            relayHub.calculateCharge(maxPossibleGas, relayRequest.relayData) +
-            relayRequest.request.value;
-        uint256 price = uint256(priceFeed.latestAnswer());
-        uint256 decimals = uint256(priceFeed.decimals());
-        tokenPreCharge = ethMaxCharge * price / (10 ** decimals);
+    returns (uint256 tokenCharge, uint256 ethCharge) {
+        ethCharge = relayHub.calculateCharge(gasUsed, relayData);
+        tokenCharge = addPaymasterFee(_weiToToken(ethCharge, priceDivisor, priceQuote));
+    }
+
+    function _tokenToWei(uint256 amount, uint256 divisor, uint256 quote) internal pure returns(uint256) {
+        return 1e18 * amount * quote / divisor;
     }
 
+    function _weiToToken(uint256 amount, uint256 divisor, uint256 quote) internal pure returns (uint256) {
+        return amount * divisor / quote / 1e18;
+    }
     // solhint-disable-next-line no-empty-blocks
     function _verifyPaymasterData(GsnTypes.RelayRequest calldata relayRequest) internal virtual override view {}
 
-    // solhint-disable-next-line no-empty-blocks
-    function _verifyApprovalData(bytes calldata approvalData) internal virtual override view {}
+    function isTokenSupported(IERC20Metadata token) public view returns (bool) {
+        return tokensSwapData[token].priceDivisor != 0;
```

Comment:
This means it is impossible to disable any token once it has been added, I think.

Reply:
Fixed